File: src/test/java/leetcode/LC008StringToIntegerAToI.java
package leetcode;

import org.junit.Test;

import static org.junit.Assert.assertEquals;

/**
 * https://leetcode.com/problems/string-to-integer-atoi/
 */
public class LC008StringToIntegerAToI {
    public class Solution {
        public int myAtoi(String str) {
            long result = 0;
            long sign = 0;
            boolean white = true, any = false;
            for (char digit : str.toCharArray()) {
                if (white) {
                    if (Character.isWhitespace(digit)) {
                        continue;
                    } else {
                        white = false;
                    }
                }
                if (Character.isDigit(digit)) {
                    if (sign == 0) {
                        sign = 1;
                    }
                    result *= 10;
                    result += digit - '0';
                } else {
                    switch (digit) {
                        case '-':
                            if (sign == 0) {
                                sign = -1;
                                continue;
                            } else {
                                return 0;
                            }
                        case '+':
                            if (sign == 0) {
                                sign = 1;
                                continue;
                            } else {
                                return 0;
                            }
                    }
                    if (any) {
                        break;
                    } else {
                        return 0;
                    }
                }
                any = true;
                if (sign * result >= Integer.MAX_VALUE) {
                    return Integer.MAX_VALUE;
                }
                if (sign * result <= Integer.MIN_VALUE) {
                    return Integer.MIN_VALUE;
                }
            }
            return (int) (sign * result);
        }
    }

    @Test
    public void test_MinusPlus3241() throws Exception {
        assertEquals(0, new Solution().myAtoi("-+3241"));
    }

    @Test
    public void test_Minus3241() throws Exception {
        assertEquals(-3241, new Solution().myAtoi("-3241"));
    }

    @Test
    public void test_SpaceMinusPlus3241A() throws Exception {
        assertEquals(-3241, new Solution().myAtoi(" -3241a"));
    }

    @Test
    public void test_9223372036854775809() throws Exception {
        assertEquals(Integer.MAX_VALUE, new Solution().myAtoi("9223372036854775809"));
    }
}

File: algorithm-design/leetcode/lc008_string_to_integer_a_to_i/lc008_string_to_integer_a_to_i.py
#!/usr/bin/env python3
# https://leetcode.com/problems/string-to-integer-atoi/

import unittest


class Solution:
    def myAtoi(self, s: str) -> int:
        result = 0
        sign = 0
        index = 0
        while index < len(s):
            if not s[index].isspace():
                break
            index += 1
        if index < len(s):
            sign = -1 if s[index] == '-' else 1
            if s[index] in ['+', '-']:
                index += 1
        while index < len(s) and s[index].isdigit():
            result *= 10
            result += ord(s[index]) - ord('0')
            if result > 2**31-1:
                break
            index += 1
        result *= sign
        return max(-2**31, min(2**31-1, result))


class TestCode(unittest.TestCase):
    def test_MinusMinus3241(self):
        self.assertEqual(Solution().myAtoi('--3241'), 0)

    def test_PlusPlus3241(self):
        self.assertEqual(Solution().myAtoi('++3241'), 0)

    def test_MinusPlus3241(self):
        self.assertEqual(Solution().myAtoi('-+3241'), 0)

    def test_Minus3241(self):
        self.assertEqual(Solution().myAtoi('-3241'), -3241)

    def test_SpaceMinusPlus3241A(self):
        self.assertEqual(Solution().myAtoi(' -3241a'), -3241)

    def test_9223372036854775809(self):
        self.assertEqual(Solution().myAtoi('9223372036854775809'), 2**31-1)

    def test_Minus9223372036854775809(self):
        self.assertEqual(Solution().myAtoi('-9223372036854775809'), -2**31)

    def test_Nothing(self):
        self.assertEqual(Solution().myAtoi('nothing'), 0)
